### PR TITLE
Added multitoken training for textual inversion. Issue 369

### PR DESCRIPTION
**Pull request by isamu-isozaki**
I used multiple tokens to represent a concept by adding num_vec_per_token number of tokens to the tokenizer which can be initialized with the initial_token. The tokens would be labeled as placeholder_token_i where i is the ith token. This does make it more verbose but it does accomplish the goal. An alternative is doing this in an inherited tokenizer class.

I also added multiple ways to initialize the tokens. For example, if the initialize_rest_random is set to true, if you have an initial token of "white dog", it'll just set all the tokens before it to random noise. But the default is for the initial tokens to be assigned in a cyclic way. For example, if num_vec_per_token is 4 and initial_token is "white dog". The weights would initially be "white dog white dog".

### CODE REVIEW

**Comment by isamu-isozaki** [comment]
@patil-suraj Sounds good! I'll also fix the conflict now. And let me know if you want me to work on the tokenizer approach like the original implementation. I have a guess on how to do it!

**Comment by patrickvonplaten** [comment]
@patil-suraj could you take a look here? :-)

**Comment by patrickvonplaten** [comment]
cc @patil-suraj here

**Comment by keturn** [comment]
And then how do we use multitoken embeddings during inference? Do we have to explicitly prompt with "a holiday card with <frida0> <frida1> <frida2>", or is there a way to just prompt "with <frida>" and the tokenizer unpacks that to its three tokens?

**Comment by isamu-isozaki** [comment]
@keturn Thanks for the q! For now, I'm doing the "a holiday card with {frida0} {frida1} {frida2}" approach but one thing I found for the original implementation is that you can concatenate the embedding of the tokens into a single vector. So, for example, it can be used like that for automatic111's API. 

But to do this, I think the tokenizer for the text encoder in the transformers library might need some changes for that. Will try it out!

**Comment by isamu-isozaki** [comment]
So, for example,
```
learned_embeds = accelerator.unwrap_model(text_encoder).get_input_embeddings().weight[placeholder_token_ids]
learned_embeds_dict = {}
learned_embeds_dict[args.placeholder_token] = learned_embeds.detach().cpu()
torch.save(learned_embeds_dict, os.path.join(args.output_dir, f"learned_embeds_{global_step}.bin"))
```
can be a file you can use for the latent-diffusion/automatic111 repo.

File: examples/textual_inversion/textual_inversion.py [hunk 1]
```diff
@@ -544,8 +569,10 @@ def main():
                 else:
                     grads = text_encoder.get_input_embeddings().weight.grad
                 # Get the index for tokens that we want to zero the grads for
-                index_grads_to_zero = torch.arange(len(tokenizer)) != placeholder_token_id
-                grads.data[index_grads_to_zero, :] = grads.data[index_grads_to_zero, :].fill_(0)
+                grad_mask = torch.arange(len(tokenizer)) != placeholder_token_ids[0]
+                for i in range(1, len(placeholder_token_ids)):
+                    grad_mask = grad_mask & (torch.arange(len(tokenizer)) != placeholder_token_ids[i])
+                grads.data[grad_mask, :] = grads.data[grad_mask, :].fill_(0)
```
Comment (Birch-san) [inline examples/textual_inversion/textual_inversion.py:575]: this masking logic has changed a bit as of https://github.com/huggingface/diffusers/pull/1665/files

I think you'll need something like:

```diff
  # Let's make sure we don't update any embedding weights besides the newly added token
- index_no_updates = torch.arange(len(tokenizer)) != placeholder_token_id
+ index_no_updates = torch.arange(len(tokenizer)) < placeholder_token_ids[0]
```

that's one of the changes I made when trying to manually rebase your changes onto my copy of the TI trainer from master:  
https://github.com/Birch-san/diffusers-play/commit/1d47955a3d827cb1de660287788c6719ede90ce5

Comment (isamu-isozaki) [reply examples/textual_inversion/textual_inversion.py:575]: nice! Thanks for the comment will update and test in a bit!

**Comment by isamu-isozaki** [comment]
Hello, while working on a personal project I think I have an idea of the tokenizer by inheriting now so I'll try pushing that today or tomorrow

**Comment by isamu-isozaki** [comment]
Hi. I added a new multitokenCLIPTokenizer which was inspired by @keturn 's suggestion of switching between <concept> and <concept>_0 <concept>_1... It's still a work in progress so I'll test it first.
I also changed refactored a code a bit

**Comment by isamu-isozaki** [comment]
ok. Currently a bit buggy but fixing.

**Comment by isamu-isozaki** [comment]
Ok! Fixed bug I'll post some results here once I'm done with training some examples

**Comment by pkurz3nd** [comment]
Hi, could this be used to load automatic1111 embeddings to diffusers sd model?
I was trying to achieve this with an extra tokenizer class similar to multitokenCLIPTokenizer and adding the loaded embeddings
to the text_encoder. However the results are nothing like the previews from the textual inversion embedding i loaded.
Am i missing something?

**Comment by isamu-isozaki** [comment]
@pkurz3nd interesting. Can you show some examples? I heard one issue is that in automatic111 you might use a different vae like [anything-v3-vae](https://huggingface.co/ckpt/anything-v3-vae-swapped). I'll try adding a load script today if you want to test it out!

**Comment by pkurz3nd** [comment]
Hi, thanks for the quick reply.
here is my notebook. 
https://colab.research.google.com/drive/1faGxmnfs3kARqa7GTmGiWKsqd-cPAE1v?usp=sharing

this is  the embedding i was testing with:
https://huggingface.co/spaablauw/FloralMarble

when i load it in torch, it says:
'sd_checkpoint': '4bdfc29c', 'sd_checkpoint_name': 'SD2.1-768'

it would be awesome if you could check out the script, i am also keep trying.

**Comment by isamu-isozaki** [comment]
sounds good lemme check

**Comment by isamu-isozaki** [comment]
@pkurz3nd Ok so checked your code. This might not be the sole issue but I don't think you assign the loaded embedding matrix to the tokenizer embeddings. So basically the tokenizer would prob just have random embeddings. Will make a load method and test it out on my end!

**Comment by pkurz3nd** [comment]
sry you are right, i somehow stopped writing on the very last line haha

this is the updated notebook
https://colab.research.google.com/drive/1-U0IdnYDGI_MJV6uC89VVhtTN51F0S_L?usp=sharing

works like a charm now with the stabilityai/stable-diffusion-2-1-base version
index

with the stabilityai/stable-diffusion-2-1 version i still get black images but thats a different problem

thanks for looking into it

**Comment by isamu-isozaki** [comment]
No worries! Looks awesome btw. For a blank image, it might be either a safety checker problem which u can disable by setting safety_checker=None 
or it might be an fp16 problem but I don't think you are using fp16.

**Comment by isamu-isozaki** [comment]
I think I recreated what you did in this (colab notebook)[https://colab.research.google.com/drive/1bt0n4gAk0DZAPcAcsdO20iAtO3td8Y6f?usp=sharing] tho I'm not able to get as good results as you

**Comment by isamu-isozaki** [comment]
Found a bit of a bug for training. Fixed and retraining now. I'll change this pr to WIP for a bit while I confirm there aren't any more bugs

**Comment by pkurz3nd** [comment]
i was running your notebook, i got an error that text in tokenizer.encode(...) is None,
and think there is a bug in the very last line of the load_multitoken_tokenizer funtion

```
        for i, placeholder_token_id in enumerate(placeholder_token_ids):
            token_embeds[placeholder_token_id] =  token_embeds[i] 
```

**Comment by isamu-isozaki** [comment]
thanks for the bug check! will fix in a bit

**Comment by isamu-isozaki** [comment]
@pkurz3nd should be fixed!
download (5)

**Comment by isamu-isozaki** [comment]
Currently working on a weird bug where regardless of what initializer token I set for training, I get various variations of this lady.
media_images_samples_3_fa8340a9070dfef796f3
this has never happened before but I thought it would be pretty interesting so I put it here.

**Comment by keturn** [comment]
That's Frida Kahlo, which makes some sense if you're trying to train the name Frida.

**Comment by isamu-isozaki** [comment]
@keturn Ohhhh so the tokenizer isn't replacing the texts properly. Thanks a bunch

**Comment by isamu-isozaki** [comment]
The issue turned out to be that I was overloading the encode method but not the __call__ method. Will fix in a bit!

**Comment by isamu-isozaki** [comment]
Ok! Should be fixed. I'll start writing some tests on the hugging face test for cleaner code.

**Comment by isamu-isozaki** [comment]
While I'm doing this pr, I think I'll add progressive tokens from latent-diffusion. The basic idea is, if I understand correctly, you first train on one token then you add another and another, etc. I am curious how it'll do compared to training from scratch on say 10 tokens.

**Comment by isamu-isozaki** [comment]
Results:

1 token training. Converged slowest but decent result

samples_11226_5058296a755e9232defb

10 token training. Converged very fast but can be a bit creepy

media_images_samples_9622_38e75da88b1b689147df

10 tokens with vector shuffling. Same as 10 tokens but more consistent

media_images_samples_10424_713eb58da7e81bce34b8

**Comment by isamu-isozaki** [comment]
Overall, I think this pr is almost finished so might open to reviews after some tests and a few features

**Comment by patrickvonplaten** [comment]
Gentle ping here @patil-suraj .

Also cc @williamberman

**Comment by patrickvonplaten** [comment]
@patil-suraj can you please give this another review

**Comment by VladAndronik** [comment]
@isamu-isozaki Thank you for your work, could you also please provide the script for inference on auto1111 embeddings?

**Comment by isamu-isozaki** [comment]
@VladAndronik Will do! @pkurz3nd gave me some hints on how to do that so should be pretty easy.

**Comment by isamu-isozaki** [comment]
@patrickvonplaten @patil-suraj  Also no pressure. I still haven't written the tests/the progressive token part yet but will try by the end of the day!

**Comment by isamu-isozaki** [comment]
I tried adding progressive tokens. I'll test if it works and if it's viable.

**Comment by isamu-isozaki** [comment]
Testing now

**Comment by isamu-isozaki** [comment]
@VladAndronik Hi! I added a function load_multitoken_tokenizer_from_automatic. The main way you can use it is, given a tensor called automatic_dict,
```
from diffusers import StableDiffusionPipeline
from textual_inversion import load_multitoken_tokenizer_from_automatic
from transformers import CLIPTextModel
from multi_token_clip import MultiTokenCLIPTokenizer
text_encoder = CLIPTextModel.from_pretrained("CompVis/stable-diffusion-v1-4", subfolder="text_encoder", revision="fp16")
tokenizer = MultiTokenCLIPTokenizer.from_pretrained("CompVis/stable-diffusion-v1-4", subfolder="tokenizer")
load_multitoken_tokenizer_from_automatic(tokenizer, text_encoder, automatic_dict, "<frida>")
vae = AutoencoderKL.from_pretrained("CompVis/stable-diffusion-v1-4", subfolder="vae", revision="fp16")
unet = UNet2DConditionModel.from_pretrained("CompVis/stable-diffusion-v1-4", subfolder="unet", revision="fp16")
pipeline =StableDiffusionPipeline(
        text_encoder=text_encoder,
        vae=vae,
        unet=unet,
        tokenizer=tokenizer,
        scheduler=PNDMScheduler(
            beta_start=0.00085, beta_end=0.012, beta_schedule="scaled_linear", skip_prk_steps=True
        ),
        safety_checker=None,
        feature_extractor=None
)
with torch.autocast("cuda"):
    image = pipeline(
                cond_embeddings, height=args.resolution, width=args.resolution, num_inference_steps=50, guidance_scale=args.guidance_scale, uncond_embeddings=uncond_embeddings
            ).images[0]
image.save('output.png')
```
I think it should generate the image you want. But let me test this first.

**Comment by isamu-isozaki** [comment]
@VladAndronik If I run above script form automatic embeddings, I get this
output

@pkurz3nd Haha finally got it working! Sorry bout the wait.

**Comment by isamu-isozaki** [comment]
For doing tests on MultiTokenCLIPTokenizer, it can't be imported into the test directory from examples/textual_inversion.py so is there a place for the class?

Usually, tokenizers and the like are part of the huggingface/transformers library but is there a preference? I can also talk about what tests I did so far.

**Comment by isamu-isozaki** [comment]
Ok. For now, I think I did everything I can do. The quality checks are for files that this pr hasn't touched and I think the same for onnx.
I still want to do some tests but @patrickvonplaten do you have some suggestions on where to put the MultiTokenTokenizer for doing tests? Currently, it's in example/textual_inversion/multi_token_clip.py

**Comment by isamu-isozaki** [comment]
Oh, and for the progressive tokens. The training run can be seen here https://wandb.ai/isamu/huggingface_textual_inv/runs/p86dttly?workspace=user-isamu it seems to be converging pretty quickly where every time a new token is added, a new detail is added. Not sure if it's better yet.

**Comment by patrickvonplaten** [comment]
cc @patil-suraj here, can you please check?

**Comment by VladAndronik** [comment]
> @VladAndronik If I run above script form automatic embeddings, I get this

@isamu-isozaki Thank you! Will try if it works for my embeddings

**Comment by VladAndronik** [comment]
@isamu-isozaki Had a problem with the code for inference, using flora embedding.
```
Traceback (most recent call last):
  File "run_embeddings.py", line 35, in <module>
    image = pipeline(
  File "/usr/local/lib/python3.8/dist-packages/torch/autograd/grad_mode.py", line 27, in decorate_context
    return func(*args, **kwargs)
  File "/usr/local/lib/python3.8/dist-packages/diffusers/pipelines/stable_diffusion/pipeline_stable_diffusion.py", line 529, in __call__
    noise_pred = self.unet(latent_model_input, t, encoder_hidden_states=text_embeddings).sample
  File "/usr/local/lib/python3.8/dist-packages/torch/nn/modules/module.py", line 1130, in _call_impl
    return forward_call(*input, **kwargs)
  File "/usr/local/lib/python3.8/dist-packages/accelerate/hooks.py", line 156, in new_forward
    output = old_forward(*args, **kwargs)
  File "/usr/local/lib/python3.8/dist-packages/diffusers/models/unet_2d_condition.py", line 418, in forward
    sample = self.conv_in(sample)
  File "/usr/local/lib/python3.8/dist-packages/torch/nn/modules/module.py", line 1130, in _call_impl
    return forward_call(*input, **kwargs)
  File "/usr/local/lib/python3.8/dist-packages/accelerate/hooks.py", line 156, in new_forward
    output = old_forward(*args, **kwargs)
  File "/usr/local/lib/python3.8/dist-packages/torch/nn/modules/conv.py", line 457, in forward
    return self._conv_forward(input, self.weight, self.bias)
  File "/usr/local/lib/python3.8/dist-packages/torch/nn/modules/conv.py", line 453, in _conv_forward
    return F.conv2d(input, weight, bias, self.stride,
RuntimeError: Input type (torch.cuda.FloatTensor) and weight type (torch.cuda.HalfTensor) should be the same
```

**Comment by isamu-isozaki** [comment]
@VladAndronik Hi. Can you show the code for run_embeddings? I think one of your models or precision might be fp32 while the rest of the models are in half precision. You can change this using .to(dtype=torch.float32) or changing the input to half precision.

**Comment by isamu-isozaki** [comment]
But the best way tends to be like having the user specify by like a mixed_precision argument and then putting everything in that precision

**Comment by VladAndronik** [comment]
@isamu-isozaki Used the code from above:
```
from diffusers import StableDiffusionPipeline, AutoencoderKL, PNDMScheduler, UNet2DConditionModel
import torch
from textual_inversion import load_multitoken_tokenizer_from_automatic
from transformers import CLIPTextModel
from multi_token_clip import MultiTokenCLIPTokenizer
automatic_dict = torch.load("automatic_embed/flora.pt")
placeholder_token="<flora>"
model = "stabilityai/stable-diffusion-2-1-base"

text_encoder = CLIPTextModel.from_pretrained(model, subfolder="text_encoder", revision="fp16").to("cuda")
tokenizer = MultiTokenCLIPTokenizer.from_pretrained(model, subfolder="tokenizer")
load_multitoken_tokenizer_from_automatic(tokenizer, text_encoder, automatic_dict, placeholder_token)
vae = AutoencoderKL.from_pretrained(model , subfolder="vae", revision="fp16").to("cuda")
unet = UNet2DConditionModel.from_pretrained(model , subfolder="unet", revision="fp16").to("cuda")
pipeline =StableDiffusionPipeline(
        text_encoder=text_encoder,
        vae=vae,
        unet=unet,
        tokenizer=tokenizer,
        scheduler=PNDMScheduler(
            beta_start=0.00085, beta_end=0.012, beta_schedule="scaled_linear", skip_prk_steps=True
        ),
        safety_checker=None,
        feature_extractor=None
)
pipeline.enable_xformers_memory_efficient_attention()
pipeline.enable_attention_slicing()
pipeline.enable_sequential_cpu_offload()
pipeline.enable_vae_slicing()
prompt =f"a photo of a {placeholder_token} cat"
image = pipeline(
            prompt, height=768, width=768, num_inference_steps=50, guidance_scale=7.5
        ).images[0]
image.show()

```

**Comment by isamu-isozaki** [comment]
@VladAndronik Oh, let me try reproducing that. Interesting that it worked for me but not for you.

**Comment by isamu-isozaki** [comment]
@VladAndronik sry will fix asap

**Comment by isamu-isozaki** [comment]
@VladAndronik I added pushed code to convert the embeddings to the text_encoder's dtype when loading. Let me know if that works.

**Comment by isamu-isozaki** [comment]
Will try fixing quality checks

**Comment by isamu-isozaki** [comment]
quality should hopefully be fixed. The issue I had was that my black module was in a lower version. For any other contributors, I recommend updating black.

**Comment by patrickvonplaten** [comment]
cc @patil-suraj could you take a look here?

**Comment by keturn** [comment]
Yes Stalebot, we still would very much like this feature.

**Comment by patrickvonplaten** [comment]
Hey @isamu-isozaki,

Super sorry to answer so late here. I don't think we'll be able to go through this anytime soon. To get this PR in, could you maybe add the new training script as a copy of the `textual_inversion` script to: https://github.com/huggingface/diffusers/tree/main/examples/research_projects? 

This way, we could merge it right away and the community can maintain it :-) Would this be ok for you?

**Comment by isamu-isozaki** [comment]
@patrickvonplaten Moved everything! And I just want to say there's no rush on my end and I'm happy to wait more + make it more minimal for an easy review!
I think a lot of the changes here came from my MultitokenTokenizer classes and such but I'm happy to remove them and go back to the original "frida" -> "frida_0", "frida_1" etc for an easy review!

File: examples/research_projects/mulit_token_textual_inversion/README.md [hunk 2]
```diff
@@ -0,0 +1,139 @@
+## This project
```
Comment (patrickvonplaten) [inline examples/research_projects/mulit_token_textual_inversion/README.md:1]:
```suggestion
## Multi Token Textual Inversion
```

File: examples/research_projects/mulit_token_textual_inversion/README.md [hunk 3]
```diff
@@ -0,0 +1,139 @@
...
+
```
Comment (patrickvonplaten) [inline examples/research_projects/mulit_token_textual_inversion/README.md:2]: The author of this project is [Isamu Isozaki](https://github.com/isamu-isozaki) - please make sure to tag the author for issue and PRs as well as @patrickvonplaten.

**Review by patrickvonplaten — APPROVED** [review]
Thanks!

**Comment by isamu-isozaki** [comment]
Should be done!

**Comment by patrickvonplaten** [comment]
Thanks a mille @isamu-isozaki ! 

Merging, FYI @patil-suraj @pcuenca @williamberman

**Comment by isamu-isozaki** [comment]
np! happy to help

**Pull request merged**